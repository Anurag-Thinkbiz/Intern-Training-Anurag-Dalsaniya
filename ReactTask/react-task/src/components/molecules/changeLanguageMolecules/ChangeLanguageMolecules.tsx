import { useTranslation } from "react-i18next";
import { languages } from "../../../utils/Constants/languageConstant";
import { useAppDispatch, useAppSelector } from "../../../data/modal/types/hookTypes/hookType";
import { changeLanguage } from "../../../redux/slices/language.slice";
import SelectMenu from "../../atoms/SelectMenu/SelectMenu";

export const ChangeLanguageMolecules = () => {
  const { i18n } = useTranslation();
  const dispatch = useAppDispatch();
  const currentLanguageFromStore = useAppSelector((state)=>state.language.language)
  
  const handleLanguageChange = (e: React.ChangeEvent<HTMLSelectElement>) => {
    i18n.changeLanguage(e.target.value);
    dispatch(changeLanguage(e.target.value));
  };
  return (
    <>
      <SelectMenu 
        options={languages}
        value={currentLanguageFromStore}
        onChange={handleLanguageChange}
      />
    </>
  );
};
