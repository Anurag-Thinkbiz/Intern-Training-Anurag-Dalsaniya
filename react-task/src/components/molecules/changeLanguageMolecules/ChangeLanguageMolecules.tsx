import { useTranslation } from "react-i18next";
import SelectMenu from "../../atoms/SelectMenu/SelectMenu";
import { languages } from "../../../utils/Constants/languageConstant";
import { useAppDispatch } from "../../../data/modal/types/hookTypes/reduxType";
import { changeLanguage } from "../../../redux/slices/language.slice";
import { useAppSelector } from "../../../data/modal/types/hookTypes/reduxType";

export const ChangeLanguageMolecules = () => {
  const { i18n } = useTranslation();
  const dispatch = useAppDispatch();
  //   const currentLanguage=localStorage.getItem('i18nextLng');
  const currentLanguage = useAppSelector((state) => state.language.language);

  const handleLanguageChange = 
    (e: React.ChangeEvent<HTMLSelectElement>) => {
      i18n.changeLanguage(e.target.value);
      dispatch(changeLanguage(e.target.value));
    }  
 

  return (
    <>
      <SelectMenu
        options={languages}
        value={currentLanguage}
        onChange={handleLanguageChange}
      />
    </>
  );
};
