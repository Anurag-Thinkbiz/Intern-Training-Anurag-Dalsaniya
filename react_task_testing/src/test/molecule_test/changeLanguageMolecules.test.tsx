import { render, screen, fireEvent } from "@testing-library/react";
import { Provider } from "react-redux";
import { configureStore, combineReducers } from "@reduxjs/toolkit";
import { PersistGate } from "redux-persist/integration/react";
import storage from "redux-persist/lib/storage";
import { persistReducer, persistStore } from "redux-persist";
import languageReducer from "../../redux/slices/language.slice";
import { ChangeLanguageMolecules } from "../../components/molecules/changeLanguageMolecules/ChangeLanguageMolecules";
import "@testing-library/jest-dom";

jest.mock("react-i18next", () => ({
  useTranslation: () => ({
    i18n: { changeLanguage: jest.fn() },
  }),
}));

const mockDispatch = jest.fn();
jest.mock("../../data/modal/types/hookTypes/hookType", () => ({
  useAppDispatch: () => mockDispatch,
  useAppSelector: jest.fn((selector) =>
    selector({ language: { language: "en" } })
  ),
}));

const persistConfig = {
  key: "root",
  storage,
};

const rootReducer = combineReducers({
  language: persistReducer(persistConfig, languageReducer),
});

const middleware = (getDefaultMiddleware: any) => {
  return getDefaultMiddleware({
    serializableCheck: false,
  });
};

const store = configureStore({ reducer: rootReducer, middleware: middleware });
const persistor = persistStore(store);

describe("ChangeLanguageMolecules with Redux Persist", () => {
  beforeEach(() => {
    mockDispatch.mockClear();
  });

  test("renders language select dropdown", () => {
    render(
      <Provider store={store}>
        <PersistGate persistor={persistor}>
          <ChangeLanguageMolecules />
        </PersistGate>
      </Provider>
    );

    const selectElement = screen.getByRole("combobox");
    expect(selectElement).toBeInTheDocument();
    expect(selectElement).toHaveValue("en");
  });

  test("updates language when a new option is selected", () => {
    render(
      <Provider store={store}>
        <PersistGate persistor={persistor}>
          <ChangeLanguageMolecules />
        </PersistGate>
      </Provider>
    );

    const selectElement = screen.getByRole("combobox");
    fireEvent.change(selectElement, { target: { value: "hi" } });

    expect(mockDispatch).toHaveBeenCalledWith({
      type: "language/changeLanguage",
      payload: "hi",
    });
  });
});
