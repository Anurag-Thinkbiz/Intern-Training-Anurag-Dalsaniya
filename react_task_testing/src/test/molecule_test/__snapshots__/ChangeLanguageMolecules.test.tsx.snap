// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChangeLanguageMolecules matches snapshot 1`] = `null`;
